tools/close_by_signature.py

```diff
@@ -0,0 +1,243 @@
+        if root_issue is None:
+            msgs_to_search_for.append(original_msg)
+            continue
+
+        msgs_to_root_issue[msg] = root_issue
+        msgs_to_search_for.append(msg)
+
+    for issue_data in issues_data_generator:
+        for msg in msgs_to_search_for:
+            if msg not in issue_data.comment.body:
+                continue
+
+            issue_data.root_issue = msgs_to_root_issue.get(msg)
+            yield issue_data
+            break
+
+        if not msgs_to_search_for:
+            yield issue_data
+
+
+def close_and_link_issues(jira, issues_data_generator):
+    for issue_data in issues_data_generator:
+        if issue_data.root_issue:
+            link_issue_to_root_issue(
+                jira=jira,
+                issue=issue_data.issue,
+                root_issue=issue_data.root_issue,
+            )
+
+        logger.debug('Closing issue with key=%s signature=%s comment=%s',
+                     issue_data.issue.key, type(issue_data.signature),
+                     issue_data.comment.body)
+
+        if dry_run_stdout is None:
+            # set issue status to 'Done'
+            jira.transition_issue(issue_data.issue, '31')
+        else:
+            dry_run_stdout.write(f'Closed issue key={issue_data.issue.key} '
+                                 f'signature={type(issue_data.signature)} '
+                                 f'comment={issue_data.comment.body}\n')
+            dry_run_stdout.flush()
+
+
+def link_issue_to_root_issue(jira, issue, root_issue):
+    logger.info('Linking issue key=%s to root issue key=%s', issue.key,
+                root_issue.key)
+
+    if dry_run_stdout is None:
+        jira.create_issue_link(
+            type='relates to',
+            inwardIssue=issue.key,
+            outwardIssue=root_issue.key,
+        )
+    else:
+        dry_run_stdout.write(f'Linked issue key={issue.key} to root issue '
+                             f'key={root_issue.key}\n')
+
+
+def main(args):
+    config_logger(args.verbose)
+
+    logger.debug('Starting issue resolver task with args=%s',
+                 pprint.pformat(args.__dict__))
+
+    username, password = get_credentials(args.user_password, args.netrc)
+    jira = get_jira_client(username, password)
+
+    signature_types = list(map(
+        lambda x: x.lower().replace('-', '').replace('_', ''),
+        args.signature_type,
+    ))
+    signatures = get_signatures(signature_types, jira)
+
+    issues = get_issues(jira, args.issue, args.recent_issues)
+
+    issues_data_generator = issues_by_signature_generator(
+        jira=jira,
+        signatures=signatures,
+        issues=issues,
+    )
+    issues_data_generator = filter_issues_by_message_generator(
+        jira=jira,
+        issues_data_generator=issues_data_generator,
+        messages=args.message,
+    )
+
+    close_and_link_issues(jira, issues_data_generator)
+
+
+if __name__ == '__main__':
+    logging.basicConfig(level=logging.DEBUG)
```

Comment:
Please move thie entire block into a proper function, otherwise all the variables inside it are considered global variables and might get accidentally used by other functions (and the IDE will give annoying "Shadowing" warnings everywhere)